File: aoc2024/src/day13.ts
import { Grid, linesToCharGrid, toNumberGrid } from "./utils/char-grid";
import { readInput, splitOnEmptyLines } from "./utils/file-utils";
import { CardinalDirection, Point } from "./utils/point";

const lines = readInput("day13", false);

function parseButton(button: string): Point {
  const steps = button.split(": ")[1];
  const [x, y] = steps.split(", ");
  const dx = parseInt(x.split("+")[1]);
  const dy = parseInt(y.split("+")[1]);
  return new Point(dx, dy);
}
function parsePrize(button: string): Point {
  const steps = button.split(": ")[1];
  const [x, y] = steps.split(", ");
  const dx = 10000000000000 + parseInt(x.split("=")[1]);
  const dy = 10000000000000 + parseInt(y.split("=")[1]);
  //   const dx = parseInt(x.split("=")[1]);
  //   const dy = parseInt(y.split("=")[1]);
  return new Point(dx, dy);
}

function part1() {
  const aPrice = 3;
  const bPrice = 1;
  const sections = splitOnEmptyLines(lines);

  let totalCost = 0;
  let winnerCount = 0;
  sections.forEach((section) => {
    const a = parseButton(section[0]);
    const b = parseButton(section[1]);
    const prize = parsePrize(section[2]);

    console.log(`${prize} : ${a} ${b}`);

    let minCost: number;

    // (a.x * na) + (b.x * nb) = prize.x
    // (a.y * na) + (b.y * nb) = prize.y
    for (let na = 0; na < 100; na++) {
      // nb = (prize.x - (a.x * na)) / b.x
      const bx = prize.x - a.x * na;
      if (bx % b.x === 0) {
        const nb = bx / b.x;
        if (nb * b.y === prize.y - a.y * na) {
          // we have a winner!
          const cost = aPrice * na + bPrice * nb;
          console.log(`Winner ${prize} with ${na}, ${nb} costs ${cost}`);
          if (!minCost || cost < minCost) {
            minCost = cost;
          }
        }
      }
    }
    if (minCost) {
      winnerCount++;
      console.log(`minCost of ${prize} is ${minCost}`);
      totalCost += minCost;
    }
  });
  console.log(`total cost: ${totalCost} for ${winnerCount} winners`);
}

function part2() {
  const aPrice = 3;
  const bPrice = 1;
  const sections = splitOnEmptyLines(lines);

  let totalCost = BigInt(0);
  sections.forEach((section) => {
    const a = parseButton(section[0]);
    const b = parseButton(section[1]);
    const prize = parsePrize(section[2]);

    console.log(`${prize} : ${a} ${b}`);

    // solve for na, if it is an integer then see if nb can be found
    // blah, should probably use matrix math for this...
    // (a.x * na) + (b.x * nb) = prize.x
    // (a.y * na) + (b.y * nb) = prize.y

    // a.y * na = prize.y - (b.y*nb)
    // na = (prize.y - (b.y * nb)) / a.y

    // b.y * nb = prize.y - (a.y * na)
    // nb = (prize.y - (a.y * na)) / b.y

    // (a.x * na) + (b.x * ((prize.y - (a.y * na)) / b.y)) = prize.x
    // (a.x * b.y * na) + (b.x * ((prize.y - (a.y * na)) = prize.x * b.y
    // (a.x * b.y * na) + (b.x * prize.y) - (b.x * a.y * na) = prize.x * b.y
    // (a.x * b.y * na) - (b.x * a.y * na) = (prize.x * b.y) - (b.x * prize.y)
    // ((a.x * b.y) - (b.x * a.y)) * na = (prize.x * b.y) - (b.x * prize.y)

    // na = ((prize.x * b.y) - (b.x * prize.y)) / ((a.x * b.y) - (b.x * a.y))

    const na = (prize.x * b.y - b.x * prize.y) / (a.x * b.y - b.x * a.y);

    if (Number.isInteger(na)) {
      const bx = prize.x - a.x * na;
      if (bx % b.x === 0) {
        const nb = bx / b.x;
        if (nb * b.y === prize.y - a.y * na) {
          // we have a winner!
          const cost = BigInt(aPrice * na) + BigInt(bPrice * nb);
          console.log(`Winner ${prize} with ${na}, ${nb} costs ${cost}`);
          totalCost += cost;
        }
      }
    }
  });
  console.log(`total cost: ${totalCost}  winners`);
}

part2();

File: aoc2024/src/python/day13.py
from collections import defaultdict
import time
from typing import Set, Tuple, List
from aoc_utils import  runIt, PuzzleInput, split_on_empty_lines, splitInts, Grid, Point, cardinal_directions, add, North, South, East, West
import functools
import math
import re
from itertools import chain, combinations, permutations, product
import sys
import numpy


testInput = r"""Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=8400, Y=5400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=12748, Y=12176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=7870, Y=6450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=18641, Y=10279"""

input = PuzzleInput('input/day13.txt', testInput)
lines = input.getInputLines(test=False)

# parse into button A, button B, prize
def parse_machine(lines:list[str]) -> tuple:
    a = tuple(map(lambda x: int(x[2:]), lines[0].split(': ')[1].split(', ')))
    b = tuple(map(lambda x: int(x[2:]), lines[1].split(': ')[1].split(', ')))
    prize = tuple(map(lambda x: int(x[2:]), lines[2].split(': ')[1].split(', ')))
    return (a, b, prize)

# I don't even...
def close_enough(a) -> bool:
    i = round(a)
    return abs(a - i) < 0.0001

def part1():
    machines = [parse_machine(s) for s in split_on_empty_lines(lines)]
    price = 0
    # print(f'found {len(machines)} machines (*4 is {len(machines) * 4})')
    for m in machines:
        # (a * m[0][0]) + (b * m[1][0]) = prize[0]
        # (a * m[0][1]) + (b * m[1][1]) = prize[1]
        A = numpy.array([[m[0][0], m[1][0]],
                         [m[0][1], m[1][1]]])
        Ai = numpy.linalg.inv(A)
        P = numpy.array([[m[2][0]], [m[2][1]]])
        r = numpy.dot(Ai, P)
        # print(r)
        if close_enough(r[0][0]) and close_enough(r[1][0]):
            price += (round(r[0][0]) * 3) + round(r[1][0])

    print(f'total price {price}')


def part2():
    machines = [parse_machine(s) for s in split_on_empty_lines(lines)]
    price = 0
    # print(f'found {len(machines)} machines (*4 is {len(machines) * 4})')
    for m in machines:
        # (a * m[0][0]) + (b * m[1][0]) = prize[0]
        # (a * m[0][1]) + (b * m[1][1]) = prize[1]
        A = numpy.array([[m[0][0], m[1][0]],
                         [m[0][1], m[1][1]]])
        P = numpy.array([[10000000000000 + m[2][0]], [10000000000000 + m[2][1]]])
        r = numpy.linalg.solve(A, P)
        # print(r)
        a = r[0][0]
        b = r[1][0]
        if close_enough(a) and close_enough(b):
            price += (round(a) * 3) + round(b)


    print(f'total price {price}')

runIt(part1, part2)
